# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def add_two(x):
    return x+2
print(add_two(3))

# lambda function:
lambda x: x+2

(lambda x: x+2)(3)


5


5

#### Saving to a Variable

In [4]:
f = lambda x, y: x+y+2
print(f(2, 6))

10


#### Multiple Inputs

In [6]:
print((lambda x, y, z: x*y*z)(3, 5, 8))

120


#### Passing a Lambda into a Function

In [7]:
def multiply(f, num):
    return f(num)
multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [16]:
def return_func():
    def multiply(num):
        return num*num
    return multiply

return_func()(5)

def return_lamb(b, c):
    return lambda x, a: x+a+b+c
return_lamb(5, 6)(5, 7)

23

#### If Statements within Lambdas

In [23]:
#Lambda x : True if (condition) else false
f = lambda num: num**2 if num >= 10 else num + 2
print(f(10))
print(f(12))

100
144


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [1]:
f = lambda x: x**3
print(f(3))

27


In [2]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [130]:
j = {
    "name": "Jesus",
    "age": 8
}
t = {
    "name": "Tate",
    "age": 18
}
m = {
    "name": "McCall",
    "age": 23
}
a = {
    "name": "Anu",
    "age": 97
}

people = [m, j, a, t]
print(sorted(people, key=lambda person_dict: person_dict["age"]))

[{'name': 'Jesus', 'age': 8}, {'name': 'Tate', 'age': 18}, {'name': 'McCall', 'age': 23}, {'name': 'Anu', 'age': 97}]


In [4]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest

[2, 4, 5, 8, 11]

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well


In [17]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4, 11, 5, 9]
more_nums = [6, 8, 7, 12]
squared_nums_map = map(squared, numbers, more_nums)
list(squared_nums_map)

[(16, 36), (11, 8), (25, 49), (9, 12)]

#### Using Lambda's with Map

In [ ]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers = [4, 9, 45, 2]
square_nums = list(map(lambda x: x**2 if x < 10 else x, numbers))
print(square_nums)

In [18]:
squared_nums2 = list(map(lambda x1, y1: (x1**2, y1**2) if x1 < 10 and y1 < 10 else (x1, y1), numbers, more_nums))
print(squared_nums2)

[(16, 36), (11, 8), (25, 49), (9, 12)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [23]:
numbers = [4, 9, 45, 2]
meth_stuff = list(map(lambda x: x * 2 - 1, numbers))
print(meth_stuff)

[7, 17, 89, 3]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [24]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Using Lambda's with Filter()

In [32]:
names = ["Jesus", "Tim", "Christina", "Eduardo", "David", "Tate"]
def t_names(name):
    if name[0].lower() == "t":
        return True
    else:
        return False
new_names_list = filter(t_names, names)
list(new_names_list)

# With lamda function
new_names_lamb = list(filter(lambda name: True if name[0].lower() == "t" else False, names))
print(new_names_lamb)

['Tim', 'Tate']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [38]:
from statistics import mean
nums = [2,7,4.2,1.6,9,4.4,4.9]
new_nums = list(filter(lambda n: True if n > mean(nums) else False, nums))
print(new_nums)

[7, 9, 4.9]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [42]:
from functools import reduce
# reduce(function,iterable)

list1 = [2, 4, 6, 8, 10]
def subtract_nums(num1, num2):
    return num1 - num2
result = reduce(subtract_nums, list1)
result

-26

#### Using Lambda's with Reduce()

In [43]:
result_lamb = reduce(lambda x, y: x-y, list1)
result_lamb

-26

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [44]:
l_5 = [1, 2, 3, 4]
multiply = reduce(lambda x, y: x*y, l_5)
print(multiply)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [51]:
def recursive(i):
    print("Hi")
    if i == 4:
        return
    i += 1
    recursive(i)
recursive(1)

Hi
Hi
Hi
Hi


#### Implementing a Base Case

In [68]:
def add_nums(num):
    # set base case
    if num <= 1:
        print("add_num(1) = 1")
        return num
    else:
        print(f"add_num({num}) = {num} + add_nums({num-1})")
        return num + add_nums(num - 1)
add_nums(10)

add_num(10) = 10 + add_nums(9)
add_num(9) = 9 + add_nums(8)
add_num(8) = 8 + add_nums(7)
add_num(7) = 7 + add_nums(6)
add_num(6) = 6 + add_nums(5)
add_num(5) = 5 + add_nums(4)
add_num(4) = 4 + add_nums(3)
add_num(3) = 3 + add_nums(2)
add_num(2) = 2 + add_nums(1)
add_num(1) = 1


55

#### Writing a Factorial Function

In [66]:
# 5! = 5*4*3*2*1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
factorial(10)

3628800

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [71]:
def rec(num):
    if num <=1:
        return 1
    else:
        return num - rec(num-1)
print(rec(5))

3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [73]:
list(range(5))
def my_range():
    for i in range(5):
        yield i
list(my_range())

[0, 1, 2, 3, 4]

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [82]:
def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start += step
list(my_range(1, 50, 1))

for i in my_range(0, 5):
    print(i)

0
1
2
3
4


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [91]:
def range1(start, stop, step=1):
    while start > stop:
        yield start**2
        start -= step
for i in range1(10, 0):
    print(i)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [9]:
places = [" ","Argentina", " ", "San Diego",""," ","","Boston","New York"]
new_list = list(filter(lambda place: True if place != " " and place != "" else False, places))
print(new_list)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [157]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
print(sorted(author, key=lambda name: name.split(" ")[-1].lower()))

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [35]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
new_temp = list(map(lambda i: f"{i[0]} {(9/5)*i[1] + 32}", places))
print(new_temp)

['Nashua 89.6', 'Boston 53.6', 'Los Angelos 111.2', 'Miami 84.2']


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [66]:
def fibonacci(i, num, num2):
    if i <= 0:
        return num
    else:
        print(num)
        j = num + num2
        num = num2
        num2 = j
        return fibonacci(i - 1, num, num2)
print(fibonacci(10, 0, 1))

0
1
1
2
3
5
8
13
21
34
55
